### PR TITLE
Add async guide

### PR DESCRIPTION
**Pull request by lmiller1990**
As discussed in #246 , some more guides for common pitfalls might be nice.

@eddyerburgh  this is just my first draft, I'd like some input on the content, style, etc. Thanks!

### CODE REVIEW

File: docs/en/guides/testing-async-components.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+describe('Foo', () => {
+  it('renders a button async when the user enters text', (done) => {
+    expect.assertions(2)
+    const wrapper = mount(Foo)
+
+    wrapper.find('input').element.value = 'Value'
+    wrapper.find('input').trigger('input')
+
+    wrapper.vm.$nextTick(() => {
+      expect(wrapper.vm.showButton).toBe(true)
```
Comment (callumacrae) [inline docs/en/guides/testing-async-components.md]: If this assertion throws, will the error be caught, or will the test time out? In my experience, it's the latter.

Comment (callumacrae) [reply docs/en/guides/testing-async-components.md]: Mind you, I've been using mocjha. might be different with jest

File: docs/en/guides/testing-async-components.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
+# Testing Asynchronous Components
```
Comment (LinusBorg) [inline docs/en/guides/testing-async-components.md]: This headline might be confused with ["Async components"](https://vuejs.org/v2/guide/components.html#Async-Components).

Maybe: "Testing Asynchronous Behaviour" ?

File: docs/en/guides/testing-async-components.md (hunk 1 above)
Comment (lmiller1990) [reply docs/en/guides/testing-async-components.md]: I think it will time out @callumacrae . What is the best thing to do in this situation?

File: docs/en/guides/testing-async-components.md (hunk 2 above)
Comment (lmiller1990) [reply docs/en/guides/testing-async-components.md]: Good catch. I think your suggested title is better. Mintor, but In my day to day life I spell it behaviour, should use use American English (no u) or English/Europe/everywhere else English?

**Comment by lmiller1990** [comment]
@eddyerburgh  good idea about mentioning `update`. What exactly does `update` do, just force `nextTick`?

File: docs/en/guides/testing-async-components.md (hunk 2 above)
Comment (LinusBorg) [reply docs/en/guides/testing-async-components.md]: Hm. I think we should probably stick do American english as that's the bigger "market"?

OTOH @eddyerburgh is british so ..., :-P

Comment (eddyerburgh) [reply docs/en/guides/testing-async-components.md]: Indeed I am British 🇬🇧

**Comment by eddyerburgh** [comment]
@lmiller1990 At the moment, it forces a component re-render.

I think we should also make it run all the component and child component watchers. I'm trying to think whether this would have unintended consequences 🤔

File: docs/en/guides/testing-async-components.md (hunk 2 above)
Comment (lmiller1990) [reply docs/en/guides/testing-async-components.md]: I am Australian (English colony) so we use the same style. But I agree we should use the American English since I believe the official docs would use that.

File: docs/en/guides/testing-async-components.md (hunk 1 above)
Comment (callumacrae) [reply docs/en/guides/testing-async-components.md]: not sure. i made an issue about it here: https://github.com/vuejs/vue-test-utils/issues/213

File: docs/en/guides/testing-async-components.md (hunk 2 above)
Comment (eddyerburgh) [reply docs/en/guides/testing-async-components.md]: Yeah, we should use US English

**Comment by lmiller1990** [comment]
Removed `watch` example, since that'll be solved in #283 .

Will look add adding more about `update`. Can you think of any good examples where running `update` is important for async tests @eddyerburgh ? I haven't had to use it much. Or were you just thinking to add something so users are more aware of how `vue-test-utils` works, in a synchronous fashion?

**Comment by 38elements** [comment]
I seem that `flushPromises` does not work for asynchronous processing executed in `Promise`.
IMHO, I think it may be is better to introduce `sleep` as below.

```js
export default function sleep (ms) {
  return new Promise(resolve => setTimeout(resolve, ms))
}
```

**Comment by eddyerburgh** [comment]
In fact, calling a synchronous method that updates state will also require calling `update`.
Any time state is changed, you need to call `update`. In all vue-test-utils methods that change or can change state, like `trigger`, `setData`, `setProps`, `update` is called to trigger a re render.

If state is changed outside of vue-test-utils, by mutating the state directly, or by triggering a method, `update` must be called to update the render tree.

@lmiller1990 An example for this guide would be when you trigger an async method by calling it on the `vm`:

```js
const Component = {
  template: `<div>{{someData}}</div>`,
  data() {
    return {
      someData: 0
    }
  },
  methods: {
    asyncMethod() {
     Promise.resolve(() => {
      this.someData++
     })
    }
  }
}

const wrapper = mount(Component)
wrapper.vm.asyncMethod()
await flushPromises()
wrapper.update()
expect(wrapper.text()).toBe(1)
```

**Comment by lmiller1990** [comment]
@eddyerburgh  I haven't actually ran into this issue - calling the methods on the wrapper seems to actually implicitly call `update` as well. The following (basically what you wrote above) actually passes.

``` js
<template>
  <div>
    {{someData}}
  </div>
</template>

<script>
  export default {
    data () {
      return {
        someData: 0
      }
    },
    methods: {
      asyncMethod () {
        Promise.resolve(() => {
        }).then(() => {
          this.someData += 1
        })
      }
    }
  }
</script>
```

``` js
import {shallow} from 'vue-test-utils'
import Update from './Update'
import flushPromises from 'flush-promises'

describe('Update', () => {
  it('works', async () => {
    const wrapper = shallow(Update)
    wrapper.vm.asyncMethod()
    await flushPromises()
    expect(wrapper.vm.someData).toBe(1)
    expect(wrapper.text()).toBe('1')
  })
})

```

I haven't actually hit any situations where I've had to call `update`. Do you have another example?

**Comment by eddyerburgh** [comment]
Ah—when flush-promises has finished, the component will have re rendered.

Looks like we don't need to include `updates` in this guide 😂

**Comment by lmiller1990** [comment]
@eddyerburgh  ok. Is there anything else you think I should change or improve? Would be nice to get this in.

Also, I fixed up the failing tests regarding `no trailing whitespace`, but it appears to be actually linting the example code, and complaining that `await is a reserved keyword` and it also doesn't like `@click`. Any way to ignore these?

**Comment by wtho** [comment]
Just a thought:. 
What if the `asyncMethod` is not promise-based, but e. g. callback-based? I know it is less often and maybe bad style, but it should still be possible to test this code.

File: docs/en/guides/testing-async-components.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+The below component makes an API call when a button is clicked, then assigns the response to `value`.
+
+``` js
+<template>
+  <button @click="fetchResults" />
+</template>
+
+<script>
+  import axios from 'axios'
```
Comment (eddyerburgh) [inline docs/en/guides/testing-async-components.md]: Could you add a line break after the import statement

File: docs/en/guides/testing-async-components.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    methods: {
+      async fetchResults () {
+        const response = await axios.get('mock/service')
+        this.value = response.data
+      }
+    }
+  }
+</script>
+```
+A test can be written like this:
```
Comment (eddyerburgh) [inline docs/en/guides/testing-async-components.md]: Can you add a link to the Jest manual mocks page—https://facebook.github.io/jest/docs/en/manual-mocks.html#content

File: docs/en/guides/testing-async-components.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+test('Foo', () => {
+  it('fetches async when a button is clicked', () => {
+    const wrapper = shallow(Foo)
+    wrapper.find('button').trigger('click')
+    expect(wrapper.vm.value).toEqual('value')
+  })
+})
+```
+
+This test currently fails, because the assertion is called before the promise resolves. One solution is to use the npm package, `flush-promises`. which immediately resolve any unresolved promises. This test is also asynchronous, so like the previous example, we need to let the test runner know to wait before making any assertions.
```
Comment (eddyerburgh) [inline docs/en/guides/testing-async-components.md]: > which immediately resolve any unresolved promises
to
> which can be used to flush all pending resolved promise handlers

File: docs/en/guides/testing-async-components.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+The updated test looks like this:
+
+``` js
+import { shallow } from 'vue-test-utils'
+import flushPromises from 'flush-promises'
+import Foo from './Foo'
+jest.mock('axios')
+
+test('Foo', () => {
+  it('fetches async when a button is clicked', () => {
```
Comment (eddyerburgh) [inline docs/en/guides/testing-async-components.md]: You need to add async to the function, that's why there's a linting error:
`it('fetches async when a button is clicked', () => {`

to

` it('fetches async when a button is clicked', async () => {`

File: docs/en/guides/testing-async-components.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    const wrapper = shallow(Foo)
+    wrapper.find('button').trigger('click')
+    expect(wrapper.vm.value).toEqual('value')
+  })
+})
+```
+
+This test currently fails, because the assertion is called before the promise resolves. One solution is to use the npm package, `flush-promises`. which immediately resolve any unresolved promises. This test is also asynchronous, so like the previous example, we need to let the test runner know to wait before making any assertions. 
+
+If you are using Jest, there are a few options, such as passing a `done` callback, as shown above. Another is to prepend the test with the ES7 'async' keyword. We can now use the the ES7 `await` keyword with `flushPromises`, to immediately resolve the API call.
```
Comment (eddyerburgh) [inline docs/en/guides/testing-async-components.md]: >such as passing a `done` callback, as shown above

It isn't shown above. Can you add an example using `done`. Instead of using flushPromises you can just use a setTimeout:

```js
test('Foo', (done) => {
   it('fetches async when a button is clicked', () => {
     const wrapper = shallow(Foo)
     wrapper.find('button').trigger('click')
     setTimeout(() => {
          expect(wrapper.vm.value).toEqual('value')
          done()
     })
   })
 })
```

(note I haven't tested this code)

We should add a note on why this works, with a link to this blog post—https://jakearchibald.com/2015/tasks-microtasks-queues-and-schedules/. Basically, the microtask queue, which processes promise callbacks, runs before the task queue, which processes timeout callbacks. So when a setTimeout callback is executed, any resolved promises will have been executed.

**Comment by lmiller1990** [comment]
Thanks for the feedback @eddyerburgh .

Updates:

1. Add example using `done`
2. Include the option to use `$nextTick` or `setTimeout` (both seem fine, I ran the tests. Do you have a preference? I used `$nextTick` in the example code, but mention both in the explanation.
3. Include link about why `setTimeout` works. I didn't know why, that was a really interesting read.
4. Improve some wording.
5. Run all tests locally (was doing so anyway, but just to check).

File: docs/en/guides/testing-async-components.md [hunk 8]
```diff
@@ -55,9 +57,24 @@ test('Foo', () => {
...
+    wrapper.find('button').trigger('click')
+    wrapper.vm.$nextTick(() => {
+      expect(wrapper.vm.value).toEqual('value')
+      done()
+    })
+  })
+})
+```
+
+The reason `$nextTick` or `setTimeout` allow the test to pass is because the microtask queue where promise callbacks are processed run efore the task queue, where `$nextTick` and `setTimeout` are processed. This means by the time the `$nexTick` and `setTimeout` run, any promise callbacks on the microtask queue will have been executed. See [here](https://jakearchibald.com/2015/tasks-microtasks-queues-and-schedules/) for a more detailed explanation.
```
Comment (eddyerburgh) [inline docs/en/guides/testing-async-components.md]: small typo — `efore`

File: docs/en/guides/testing-async-components.md [hunk 9]
```diff
@@ -55,9 +57,24 @@ test('Foo', () => {
...
 
-If you are using Jest, there are a few options, such as passing a `done` callback, as shown above. Another is to prepend the test with the ES7 'async' keyword. We can now use the the ES7 `await` keyword with `flushPromises`, to immediately resolve the API call.
+Another solution is to use the npm package, `flush-promises`, which can be used to flush all pending resolved promise handlers, and prepend the test with the ES7 'async' keyword. The ES7 `await` keyword with `flushPromises` can now be used to immediately resolve the API call.
```
Comment (eddyerburgh) [inline docs/en/guides/testing-async-components.md]: Can we reword this:
>Another solution is to use the npm package, `flush-promises`, which can be used to flush all pending resolved promise handlers, and prepend the test with the ES7 'async' keyword. The ES7 `await` keyword with `flushPromises` can now be used to immediately resolve the API call.

To something like this:
>Another solution is to use an `async` function and the npm package `flush-promises`. `flush-promises` flushes all pending resolved promise handlers. You can `await` the call of `flushPromises` to flush the promises and improve the readability of your test.

**Comment by lmiller1990** [comment]
Done! Thanks for the feedback.

**Comment by doun** [comment]
@lmiller1990 Would u add an example for catching exceptions:
`
try{
await a();
expect(CONDITION);
await b();
expect(CONDITION);
}
catch(e){
}
`
be careful that exception throw by expect() should NOT be caught, or it will pass all tests

**Comment by lmiller1990** [comment]
@doun  the current test shouldn't swallow any exceptions, but throw an error. I haven't experienced what your describe before, can you link to an example/post some sample code? Or do you mean I should add an example showing how to test if an error was throw from some async code?

**Comment by doun** [comment]
It's not about current sample code.
I just did this days ago, sample code:
```javascript
test("expect fail", async ()=>{
try{
  var resultA = await TaskA()
  expect(resultA).toBe(true)
  var resultB = await TaskB()
  expect(true).toBe(false)
}
//all exceptions caught, the test will pass
catch(e){
}
})
```
I wanted to catch only async exceptions, but the test framework(jest) also use exceptions to catch expect failures.

**Comment by lmiller1990** [comment]
@doun Thanks for the sample. It appears you are just testing regular JavaScript methods, not Vue component specific ones? Or can you provide the component and some more context?

I'm not sure this belongs here -- of course it is important, but it doesn't feel specific enough to Vue components to include.

One good practise can be to use `expect.assertions` at the top of your tests - that way, if an error occurs and Jest skips to the `catch` block, because `expect.assertions` is not the correct amount, the test will fail.

**Comment by doun** [comment]
No, it’s about async test, not about vue component test. Your idea is awesome, thanks!

File: docs/en/guides/testing-async-components.md [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+test('Foo', () => {
+  it('fetches async when a button is clicked', () => {
+    const wrapper = shallow(Foo)
+    wrapper.find('button').trigger('click')
+    expect(wrapper.vm.value).toEqual('value')
+  })
+})
+```
+
+This test currently fails because the assertion is called before the promise in `fetchResults` resolves. Most unit test libraries provide a callback to let the runner know when the test is complete. Jest and Karma both use `done`. We can use `done` in combination with `$nextTick` or `setTimeout` to ensure any promises resolve before the assertion is made.
```
Comment (eddyerburgh) [inline docs/en/guides/testing-async-components.md]: > Jest and Karma both use 

should be 

Jest and Mocha both use

File: docs/en/guides/testing-async-components.md [hunk 11]
```diff
@@ -0,0 +1,98 @@
+# Testing Asynchronous Behavior
+
+To simplify testing, `vue-test-utils` applies updates _synchronously_. However, there are some techniques you need to be aware of when testing a component with asynchronous behavior such as callbacks or promises.
```
Comment (eddyerburgh) [inline docs/en/guides/testing-async-components.md]: > applies updates _synchronously_

to

>  applies DOM updates _synchronously_

**Comment by lmiller1990** [comment]
Thanks @eddyerburgh , I added those changes.

**Comment by eddyerburgh** [comment]
Thanks @lmiller1990 🙂

**Comment by lmiller1990** [comment]
Thank you @eddyerburgh , the part about how `setTimeout` and task/microtasks was new to me. OSS is great. I feel like I've learned more doing OS in the last year than like 4 years at school at 2 in the workforce.

**Comment by Ray-56** [comment]
if
image had setTimeout 1500s, expected is error. 
how do i handle it?

**Comment by lmiller1990** [comment]
@guokangf can you post your question on stackoverflow with `vue-test-utils` tag and the test you trying to write? I can help you there. This repo is just for issues with `vue-test-utils`.

Thanks.
